TFastModbus.cpp

```diff
@@ -0,0 +1,173 @@
+            }
+        } else {
+            if ((dataLength + availableDataLength) > bufferSize) {
+                DEBUG("ERROR Fast modbus scan buffer too short\n");
+                return 0;
+            }
+            for (int i = 0; i < availableDataLength; i++) {
+                buffer[dataLength] = (uint8_t)Serial->read();
+                dataLength++;
+            }
+        }
+    }
+    return dataLength;
+}
+
+// Continues scan cycle until 0x04 end scan subcommand would reach
+bool TFastModbus::ContinueScan(uint8_t* serialNumber, uint8_t* modbusAddress)
+{
+    // Fast modbus packet with continue scan command
+    //  0xFD - address for scan, 0x60 command, 0x02 - continue subcommand
+    uint8_t continueData[] = {0xFD, 0x60, 0x02, 0x49, 0xF1};
+    if (Serial->write(continueData, sizeof(continueData)) != sizeof(continueData)) {
+        DEBUG("*** ERROR Sending fast modbus scan continue!\n");
+        return false;
+    }
+
+    uint8_t scanData[32];
+    uint8_t scanDataLength = ReadBytes(scanData, sizeof(scanData), TimeoutMs);
+    if (!scanDataLength) {
+        DEBUG("*** ERROR Reading fast modbus scan response!\n");
+        return false;
+    }
+
+    uint8_t* fastModbusPacket = scanData;
+    uint8_t fastModbusPacketLength = scanDataLength;
+    while ((fastModbusPacketLength > 0) && (*fastModbusPacket == 0xFF)) {
+        fastModbusPacket += 1;
+        fastModbusPacketLength -= 1;
+    }
+
+    DEBUG("Fast modbus packet length: ");
+    DEBUG(fastModbusPacketLength);
+    DEBUG("\n");
+    for (int i = 0; i < fastModbusPacketLength; i++) {
+        DEBUG(fastModbusPacket[i], 16);
+    }
+    DEBUG("\n");
+
+    uint16_t checkCrc = CrcClass::crc16_modbus(fastModbusPacket, fastModbusPacketLength - 2);
+    uint16_t crc = (fastModbusPacket[fastModbusPacketLength - 1] << 8) + fastModbusPacket[fastModbusPacketLength - 2];
+    if (checkCrc != crc) {
+        DEBUG("*** ERROR CRC mismatch! ");
+        DEBUG(checkCrc, 16);
+        DEBUG(" expected, but ");
+        DEBUG(crc, 16);
+        DEBUG(" actually received!\n");
+        return false;
+    }
+
+    // Check if packet has wrong address and fast modbus command
+    if (fastModbusPacket[0] != 0xFD || fastModbusPacket[1] != 0x60) {
+        DEBUG("*** ERROR Wrong modbus address or fast modbus command: 0xFD address and 0x60 command expected, but ");
+        DEBUG(fastModbusPacket[0], 16);
+        DEBUG(" and ");
+        DEBUG(fastModbusPacket[1], 16);
+        DEBUG(" actually received!\n");
+        return false;
+    }
+
+    switch (fastModbusPacket[2]) {
+        // If fast modbus subcommand==0x03 the response contain device data
+        case 0x03:
+            memcpy(serialNumber, &fastModbusPacket[3], WB_MSW_SERIAL_NUMBER_SIZE);
+            memcpy(modbusAddress, &fastModbusPacket[7], 1);
+            return true;
+        case 0x04:
+            // If fast modbus subcommand==0x04, there are no unsolicited devices left on fast modbus bus
```

Comment:
`unsolicited` - а это слово точно из нашего контекста? А то [интернеты говорят](https://context.reverso.net/%D0%BF%D0%B5%D1%80%D0%B5%D0%B2%D0%BE%D0%B4/%D0%B0%D0%BD%D0%B3%D0%BB%D0%B8%D0%B9%D1%81%D0%BA%D0%B8%D0%B9-%D1%80%D1%83%D1%81%D1%81%D0%BA%D0%B8%D0%B9/unsolicited), что оно скорее означает "нежелательный", чем "неопрошенный".

```suggestion
            // subcommand 0x04 means that scan is finished
```